### ISSUE TITLE
Specify connection config via an object and not string

### ISSUE THREAD

**Issue by ad-si** [Automattic/mongoose#10339]
Sorry if I'm blind, but how can I specify the connection with an object and not a string? E.g.

```js
mongoose.connect({
  username: 'user123',
  password: 'pass123',
  host: 'mongodb0.example.com',
  port: 1234,
})
```

**Comment by IslandRhythms** [comment]
Do you mean how is it possible to not use a string or specify the connection string with an object instead of a string?

**Comment by ad-si** [comment]
> Do you mean how is it possible to not use a string or specify the connection string with an object instead of a string?

Yeah, pretty much exactly like in my example

**Comment by ad-si** [comment]
😳  Wow, I'm baffled that this is not supported.
Well then this is a feature request for it. An Object is 
1. much easier to handle
2. Can give much better error messages in combination with TypeScript 
3. Is the JS default for configuring anything

**Comment by ShadiestGoat** [comment]
I mean to be fair though, this isn't that hard to implement. My current setup is just get the config, and generate a link using a config

```ts
const conf:{
    mongodb: {
        username: string,
        password: string,
        ip: string,
        db: string,
        port: string
    }
} = configFunction()
const MongoDBendPoint = "mongodb://" + conf.mongodb.username + ":" + conf.mongodb.password + "@" + conf.mongodb.ip + ":" + conf.mongodb.port + "/" + conf.mongodb.db + "?readPreference=primary&appname=MyAppNameHere&ssl=false?authSource=" + conf.mongodb.db
//or
const MongoDBendPoint2 = `mongodb://${conf.mongodb.username}:${conf.mongodb.password}@${conf.mongodb.ip}:${conf.mongodb.port}/${conf.mongodb.db}?readPreference=primary&appname=MyAppNameHere&ssl=false?authSource=${conf.mongodb.db}`
```

**Comment by ad-si** [comment]
1. Did you handle URL encoding correctly?
2. A lot of invalid values for the parameters could generate a potentially valid MongoDB string. Such things are quite error prone and hard to debug.
3. This code is exceptionally convoluted and ugly. More potential for mistakes. (Would be too long for the common 80 chars limit per line anyways)

What not simply add the function to build this string to mongoose? Including type / range checking, normalization, and so on.

**Comment by ShadiestGoat** [comment]
I'm not saying it's a perfect solution, I'm just saying that it could be a workaround-ish solution, for now. I'm very not against having this function be in mongoose

**Comment by vkarpov15** [comment]
Not a bad idea. The only potential downside is that connection string is MongoDB's default approach, so makes it harder to copy/paste connection string from Atlas or Studio 3T. But worth considering for the future.

**Comment by ShadiestGoat** [comment]
Wait, are you suggesting to replace it completely?

**Comment by ad-si** [comment]
Well I don't 😳

**Comment by vkarpov15** [comment]
@ShadiestGoat not suggesting replacing it completely, that would be too big a breaking change for too little benefit IMO. But a [named parameters](https://masteringjs.io/tutorials/fundamentals/parameters) style syntax for `connect()` might be a good new feature, especially given that MongoDB expects you to escape username/password before putting them into connection strings.

**Comment by ShadiestGoat** [comment]
Yeah, I no sorry I was a bit worried that you were going to replace it completely, I don't see any benefits to that personally. The connect should be `string | loginParams`

**Comment by ad-si** [comment]
Actually, on second thought, I think it would be cleaner to have two functions:
```
mongoose.connect :: string -> Connection
mongoose.connectConfig :: loginParams -> Connection
```

**Comment by vkarpov15** [comment]
Sure, `mongoose.connectWithConfig()` and `Connection#openWithConfig()` sounds like a reasonable alternative to avoid too many different overrides for `mongoose.connect()`.
